File: src/main/java/com/github/kayjamlang/core/provider/ExpressionProvider.java
package com.github.kayjamlang.core.provider;

import com.github.kayjamlang.core.Type;
import com.github.kayjamlang.core.exceptions.TypeException;

public abstract class ExpressionProvider<ET, RO, CO, MCO> {
    /**
     * @deprecated please replace to {@code mainProvider.defaultObject}
     */
    @Deprecated
    public RO provide(MainExpressionProvider<RO, CO, MCO> mainProvider, CO context, CO argsContext, ET expression) {
        return mainProvider.defaultObject;
    }

    /**
     * @deprecated please don't use this
     */
    @Deprecated
    public Type getType(MainExpressionProvider<RO, CO, MCO> mainProvider, CO context, CO argsContext, ET expression) throws TypeException {
        return Type.ANY;
    }
}

File: src/main/scala/com/github/kayjamlang/core/provider/ExpressionProvider.scala
package com.github.kayjamlang.core.provider

import com.github.kayjamlang.core.Type
import com.github.kayjamlang.core.exception.TypeException
import com.github.kayjamlang.core.expression.Expression
import com.github.kayjamlang.core.opcodes.AccessType

class ExpressionProvider[A, B, C, D](accessType: AccessType) extends Expression(accessType, -1) {
  def provide(mainProvider: MainExpressionProvider[B, C, D], context: C, argsContext: C, expression: A): B = mainProvider.defaultObject

  @throws[TypeException]
  def getType(mainProvider: MainExpressionProvider[B, C, D], context: C, argsContext: C, expression: A): Type = Type ANY
}
